ublox/src/parser.rs

```diff
@@ -0,0 +1,121 @@
+use crate::{
+    error::ParserError,
+    ubx_packets::{match_packet, ubx_checksum, PacketRef, SYNC_CHAR_1, SYNC_CHAR_2},
+};
+
+/// Some big number, TODO: need validation against all known packets
+const MAX_PACK_LEN: usize = 1022;
+
+/// Streaming parser for UBX protocol with buffer
+#[derive(Default)]
+pub struct Parser {
+    buf: Vec<u8>,
+}
+
+impl Parser {
+    pub fn is_buffer_empty(&self) -> bool {
+        self.buf.is_empty()
+    }
+    pub fn buffer_len(&self) -> usize {
+        self.buf.len()
+    }
+    pub fn consume<'a, 'b, 'c>(&'a mut self, new_data: &'b [u8]) -> ParserIter<'c>
+    where
+        'a: 'c,
+    {
+        match self
+            .buf
+            .iter()
+            .chain(new_data.iter())
+            .position(|x| *x == SYNC_CHAR_1)
+        {
+            Some(mut off) => {
+                if off >= self.buf.len() {
+                    off -= self.buf.len();
+                    self.buf.clear();
+                    self.buf.extend_from_slice(&new_data[off..]);
+                    off = 0;
+                } else {
+                    self.buf.extend_from_slice(new_data);
+                }
+                ParserIter {
+                    buf: &mut self.buf,
+                    off,
+                }
+            }
+            None => {
+                self.buf.clear();
+                ParserIter {
+                    buf: &mut self.buf,
+                    off: 0,
+                }
+            }
+        }
+    }
+}
+
+/// Iterator over data stored in `Parser` buffer
+pub struct ParserIter<'a> {
+    buf: &'a mut Vec<u8>,
+    off: usize,
+}
+
+impl<'a> Drop for ParserIter<'a> {
+    fn drop(&mut self) {
+        if self.off <= self.buf.len() {
+            self.buf.drain(0..self.off);
+        }
+    }
+}
+
+impl<'a> ParserIter<'a> {
+    /// Analog of `core::iter::Iterator::next`, should be switched to
+    /// trait implmentation after merge of https://github.com/rust-lang/rust/issues/44265
+    pub fn next(&mut self) -> Option<Result<PacketRef, ParserError>> {
+        while self.off < self.buf.len() {
+            let data = &self.buf[self.off..];
+            let pos = match data.iter().position(|x| *x == SYNC_CHAR_1) {
```

Comment:
Modifying `self.off` is bad idea, but it is good idea to create local variable with `&data[pos`

Reply:
Why is modifying `self.off` a bad idea? Don't we do it anyway below?

Reply:
> Why is modifying self.off a bad idea? Don't we do it anyway below?

My supposition, 99% of time we will handle continuous stream of valid packets.
In that case it would be like this (with `self.off += pos` modification as you suggest):
```
let pos = 0;//99% cases

self.off += 0; // actually `self.off += pos` , but in 99% pos == 0`
if not enough data {
   return None;
}
self.off += 6 + pack_len + 2;
```

you see, it would be the same code with `self.off +pos` at start and without if we get full packet, but if would be unnecessary
`self.off +=0` every time for partly received packet.

So if user pass buffer with length 1 to `consume` (as for example [here](https://github.com/lkolbly/ublox/blob/567263b5011a7367236de1f6f0ce5e6528d99d76/ublox/src/serialport.rs#L393) and in my own code, untill I read protocol and changed buffer length to 8) we get `28/*NavPosLlh payload*/ + 6 + 2` calls to `next` and only one of them will really update  `self.off`, while the others would only:
```
self.off += 0;
```
and compiler can not optimize this, because of it calculate `pos` every time.